### ISSUE TITLE
Error in reading CSV fields when a delimiter char present in field for string delimiter

### ISSUE THREAD

**Issue by senthilbalu** [JoshClose/CsvHelper#450]
Hi 
I want use a custom delimiter (&|$) as my text in file may contain if I have single char delimiter.
Please refer the sample code below :

````  var csvData = @"
CompanyName&|$Code&|$City
ABCDxyz Ltd&|$ABC&|$Mexico
MADD & sons&|$MAC&|$Sydney     
";````

            var stream = new MemoryStream(Encoding.UTF8.GetBytes(csvData));
            using (var reader = new CsvReader(new StreamReader(stream)))
            {
                reader.Configuration.Delimiter = "&|$";
                reader.Configuration.IgnoreQuotes = true;
                while(reader.Read())
                    Console.Out.WriteLine(reader[0] + "\t" + reader[1] + "\t" + reader[2]);             
            }  


The Output was :


````
CompanyName     Code    City
ABC Ltd 		ABC     Mexico
MADD    		ons     MAC
````
Please note that instead of reading MADD & sons as single field it breaks the field as the field contains the first delimiter char. 

Can you please look into this issue and help me with a solution?

**Comment by JoshClose** [comment]
Sounds like a bug. Do you have control over the data being output? If so, you could try escaping the field with `"`s.

**Comment by senthilbalu** [comment]
Hi Josh
Unfortunately I don't have control over them as it is sourced from a different system. Is there something u can do for this to be fixed ?

**Comment by jsnape** [comment]
Hi @JoshClose,
I work with @senthilbalu and I'm just taking a look to see if I can figure out how to fix this. We don't have github write access from our internal network so I'll paste code here if that's OK.

I have a unit test that can be added to CsvParserDelimiterTests:

```CSharp
        [TestMethod]
        public void MultipleCharDelimiter3TestWithDelimCharInData()
        {
            using (var stream = new MemoryStream())
            using (var reader = new StreamReader(stream))
            using (var writer = new StreamWriter(stream))
            using (var parser = new CsvParser(reader))
            {
                writer.WriteLine("ABCDxyz Ltd&|$ABC&|$Mexico");
                writer.WriteLine("MADD & sons&|$MAC&|$Sydney");
                writer.Flush();
                stream.Position = 0;

                parser.Configuration.Delimiter = "&|$";
                parser.Configuration.HasHeaderRecord = false;

                var row = parser.Read();
                Assert.IsNotNull(row);
                Assert.AreEqual(3, row.Length);
                Assert.AreEqual("ABCDxyz Ltd", row[0]);
                Assert.AreEqual("ABC", row[1]);
                Assert.AreEqual("Mexico", row[2]);

                row = parser.Read();
                Assert.IsNotNull(row);
                Assert.AreEqual(3, row.Length);
                Assert.AreEqual("MADD & sons", row[0]);
                Assert.AreEqual("MAC", row[1]);
                Assert.AreEqual("Sydney", row[2]);

                row = parser.Read();
                Assert.IsNull(row);
            }
        }
```

Looking at the code I think the issue is in CsvParser::ReadLine(). The code assumes that once you hit the first character of the delimiter then the rest must follow:

```CSharp
else if( c == configuration.Delimiter[0] || inDelimiter )
{
	if( !inDelimiter )
	{
		// If we hit the delimiter, we are
		// done reading the field and can
		// add it to the record.
		AppendField( ref field, fieldStartPosition, readerBufferPosition - fieldStartPosition - 1 );
		// Include the comma in the byte count.
		UpdateBytePosition( fieldStartPosition, readerBufferPosition - fieldStartPosition );
		AddFieldToRecord( ref recordPosition, field, ref fieldIsBad );
		fieldStartPosition = readerBufferPosition;
		field = null;

		inDelimiter = true;
	}
					
	if( delimiterPosition == configuration.Delimiter.Length - 1 )
	{
		// We are done reading the delimeter.

		// Include the delimiter in the byte count.
		UpdateBytePosition( fieldStartPosition, readerBufferPosition - fieldStartPosition );
		inDelimiter = false;
		prevCharWasDelimiter = true;
		delimiterPosition = 0;
		fieldStartPosition = readerBufferPosition;
	}
	else
	{
		delimiterPosition++;
	}
}

```

Just thinking about a fix... The first character of the delimiter is checked multiple times in the ReadLine() method. All of them would need to differentiate between the full delimiter and some of the chars.

I've managed to write a function that allows the test above to pass:

```CSharp
protected bool IsDelimiter(char c)
{
    if (c != configuration.Delimiter[0])
    {
        return false;
    }

    int nextPosition = readerBufferPosition;
    int delimiterIndex = 1; // Because we already tested the first char above.

    while (nextPosition < charsRead && delimiterIndex < configuration.Delimiter.Length)
    {
        if (readerBuffer[nextPosition] != configuration.Delimiter[delimiterIndex])
        {
            return false;
        }

        ++delimiterIndex;
        ++nextPosition;
    }

    return delimiterIndex == configuration.Delimiter.Length;
}
```

But this now causes a failure in MultipleCharDelimiterWithBufferEndingInMiddleOfDelimiterTest() - I suspect the answer lies in using GetChar() but with all the parameters I'm afraid I don't quite understand it.

Any guidance you can give would be appreciated.

**Comment by jsnape** [comment]
I've tried a number of different solutions but I end up breaking a number of other unit tests whichever approach I try. I may have to give up on this.

**Comment by JoshClose** [comment]
I'm going to try and look at some of these major bugs this weekend here.

**Comment by JoshClose** [comment]
I just fixed this an pushed version 2.13.3 to NuGet. There is the possibility of this not working if the buffer runs out in the middle of the delimiter. I'm in the process of re-writing the parser, and this won't be an issue then.

**Comment by jsnape** [comment]
Thanks. I've upgraded our references and we will give it a whirl.

@senthilbalu will you please test with the file that caused the issue in the first place?
